projects/cdk/directives/auto-focus/autofocus.directive.ts

```diff
@@ -32,57 +32,85 @@ export class TuiAutoFocusDirective implements AfterViewInit {
-        @Inject(INJECTOR) private readonly injector: Injector,
-        @Inject(ElementRef) private readonly elementRef: ElementRef<HTMLElement>,
         @Optional()
         @Self()
         @Inject(TUI_FOCUSABLE_ITEM_ACCESSOR)
         private readonly tuiFocusableComponent: TuiFocusableElementAccessor | null,
         @Inject(TUI_IS_IOS) private readonly isIos: boolean,
         @Inject(ANIMATION_FRAME) private readonly animationFrame$: Observable<number>,
+        @Inject(NgZone)
+        private readonly ngZone: NgZone,
+        @Inject(ElementRef)
+        private readonly elementRef: ElementRef<HTMLElement>,
+        @Inject(Renderer2)
+        private readonly renderer: Renderer2,
     ) {}
 
     ngAfterViewInit() {
         if (!this.autoFocus) {
             return;
         }
 
-        const element =
+        if (this.isIos && this.isTextFieldElement()) {
+            this.ngZone.runOutsideAngular(() => this.iosWebkitAutofocus());
+        } else {
+            race(
+                timer(TIMEOUT),
+                this.animationFrame$.pipe(
+                    throttleTime(POLLING_TIME),
+                    map(() => this.element.closest(NG_ANIMATION_SELECTOR)),
+                    skipWhile(Boolean),
+                    take(1),
+                ),
+            ).subscribe(() => setNativeFocused(this.element));
+        }
+    }
+
+    private get element(): TuiNativeFocusableElement {
+        return (
             this.tuiFocusableComponent?.nativeFocusableElement ||
-            this.elementRef.nativeElement;
+            this.elementRef.nativeElement
+        );
+    }
 
-        if (this.isIos) {
-            veryVerySadIosFix(this.injector);
-        }
+    private isTextFieldElement(): boolean {
+        return this.element.matches('input, textarea');
+    }
 
-        race(
-            timer(TIMEOUT),
-            this.animationFrame$.pipe(
-                throttleTime(POLLING_TIME),
-                map(() => element.closest(NG_ANIMATION_SELECTOR)),
-                filter(v => !v),
-                take(1),
-            ),
-        ).subscribe(() => {
-            setNativeFocused(element);
+    private iosWebkitAutofocus() {
+        const fakeInput: HTMLElement = this.renderer.createElement('input');
+
+        fakeInput.style.position = 'absolute';
+        fakeInput.style.opacity = '0';
+        fakeInput.style.height = '0';
+
+        /**
+         * @note: ping-pong eager strategy hack
+         * After creating an element and bringing it into DOM,
+         * the browser automatically focuses on the invisible element.
+         * And then, after focus is triggered, we try to focus on target element, and if we managed to refocus,
+         * then we try to focus again on an invisible element, so that the keyboard slowly appears.
+         * This ping pong allows the keyboard to not overlap the modal window.
+         */
+        fakeInput.addEventListener('blur', () => setNativeFocused(fakeInput), {
```

Comment:
Let's move adding both listeners together, one after another, so it's easier to read